File: Poj/1716.pas
Const Maxn=10000;
Const Maxm=1000000;
Var
  c,t,nt:Array[-1..Maxm]Of Longint;
  Ind,f,q:Array[-1..Maxn]Of Longint;
  Inq:Array[-1..Maxn]Of Boolean;
  i,j,k,n,m:Longint;
  x,y,z,cnt:Longint;
Procedure AE(i,j,k:Longint);
Begin
  Inc(cnt);
  t[cnt]:=j;
  c[cnt]:=k;
  nt[cnt]:=ind[i];
  ind[i]:=cnt;
End;
Function SPfa():Longint;
Var v,i,j,k,l,r:Longint;
Begin
  For i:=-1 To 10000 Do
    f[i]:=-Maxlongint;
  Fillchar(inq,Sizeof(inq),False);
  f[-1]:=0;
  l:=0;
  r:=0;
  q[l]:=-1;
  While l<=r Do Begin
    v:=q[l mod n];
    Inc(l);
    Inq[v]:=False;
    k:=Ind[v];
    While k<>-1 Do Begin
      If f[v]+c[k]>f[t[k]] Then Begin
        f[t[k]]:=f[v]+c[k];
        If Not inq[t[k]] Then Begin
          Inc(r);
          q[r mod n]:=t[k];
          Inq[t[k]]:=True;
        End;
      End;
      k:=nt[k];
    End;
  End;
  Exit(f[n]-f[-1]);
End;
Begin
  //Assign(Input,'a.in');
  Reset(Input);
  Readln(m);
  n:=0;
  cnt:=0;
  For I:=-1 To Maxn Do
    Ind[i]:=-1;
  For i:=1 To m Do Begin
    Readln(x,y);
    AE(x-1,y,2);
    If y>n Then N:=y;
  End;
  For I:=-1 To n-1 Do Begin
    AE(i,i+1,0);
    AE(i+1,i,-1);
  End;
  Writeln(spfa);
End.

File: Poj/1716.cpp
#include <cstdio>
#include <iostream>
#include <cstring>
#include <string>
#define MAXN 10005
#define MAXM 10000000

using namespace std;

int cnt;
int t[MAXM], nt[MAXM], c[MAXM], ind[MAXN];
int q[MAXN], vis[MAXN];
int f[MAXN];

void add(int a, int b, int w)
{
    cnt++;
    t[cnt] = b;
    c[cnt] = w;
    nt[cnt] = ind[a];
    ind[a] = cnt;
}

int main()
{
    int m;
    int n = 0;
    cin >> m;
    cnt = 0;
    for (int i = 0; i < 10005; i++) ind[i] = -1;
    for (int i = 0; i < m; i++)
    {
        int a, b;
        scanf("%d%d", &a, &b);
        a++;
        b++;
        if (a > b) swap(a, b);
        n = max(n, b);
        add(a - 1, b, 2);
    }
    
    for (int i = 0; i <= n; i++)
    {
        f[i] = - 2147483647;
        add(i, i + 1, 0);
        add(i + 1, i, -1);
    }
    
    f[0] = 0;
    memset(q, 0, sizeof q);
    memset(vis, 0, sizeof vis);
    
    int l = 0, r = 0;
    q[l] = 0;
    vis[0] = 1;
    while (l <= r)
    {
        int x = q[l % n];
        l++;
        vis[x] = 0;
        for (int k = ind[x]; k != -1; k = nt[k])
        {
            if (f[t[k]] < f[x] + c[k])
            {
                f[t[k]] = f[x] + c[k];
                if (!vis[t[k]])
                {
                    r++;
                    q[r % n] = t[k];
                    vis[t[k]] = 1;
                }
            }			
        }
    }
    
    cout << f[n] - f[0] << endl;
    return 0;
}